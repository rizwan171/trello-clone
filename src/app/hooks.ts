import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
import type { RootState, AppDispatch } from "./store";

// The below useAppDispatch does not require a return type, hence the eslint disable comment
// See https://stackoverflow.com/questions/70711795/missing-return-type-on-function-useappdispatch for more info
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
