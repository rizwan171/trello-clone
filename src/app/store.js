import { configureStore } from "@reduxjs/toolkit";
import boardReducer from "../features/boardSlice.js";
import listsReducer from "../features/listsSlice.js";
import cardsReducer from "../features/cardsSlice.js";
import boardOptionsReducer from "../features/boardOptionSlice.js";
import currentSelectedCardReducer from "../features/currentSelectedCardSlice.js";
import currentSelectedListReducer from "../features/currentSelectedListSlice.js";
import listOptionsMenuPositionReducer from "../features/listOptionsMenuPositionSlice.js";
import tagsReducer from "../features/tagsSlice.js";
import selectedTagColourReducer from "../features/selectedTagColourSlice.js";
import createTagMenuDataReducer from "../features/createTagMenuDataSlice.js";
import modalActionMenusVisibilityReducer from "../features/modalActionMenusVisibilitySlice.js";

export const store = configureStore({
  reducer: {
    board: boardReducer,
    lists: listsReducer,
    cards: cardsReducer,
    boardOptions: boardOptionsReducer,
    currentSelectedCard: currentSelectedCardReducer,
    currentSelectedList: currentSelectedListReducer,
    listOptionsMenuPosition: listOptionsMenuPositionReducer,
    tags: tagsReducer,
    selectedTagColour: selectedTagColourReducer,
    createTagMenuData: createTagMenuDataReducer,
    modalActionMenusVisibility: modalActionMenusVisibilityReducer,
  },
});
