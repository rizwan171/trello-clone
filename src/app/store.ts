import { combineReducers, configureStore, PreloadedState } from "@reduxjs/toolkit";
import boardReducer from "../features/boardSlice";
import listsReducer from "../features/listsSlice";
import cardsReducer from "../features/cardsSlice";
import boardOptionsReducer from "../features/boardOptionSlice";
import currentSelectedCardReducer from "../features/currentSelectedCardSlice";
import currentSelectedListReducer from "../features/currentSelectedListSlice";
import listOptionsMenuPositionReducer from "../features/listOptionsMenuPositionSlice";
import tagsReducer from "../features/tagsSlice";
import selectedTagColourReducer from "../features/selectedTagColourSlice";
import createTagMenuDataReducer from "../features/createTagMenuDataSlice";
import modalActionMenusVisibilityReducer from "../features/modalActionMenusVisibilitySlice";

const rootReducer = combineReducers({
  board: boardReducer,
  lists: listsReducer,
  cards: cardsReducer,
  boardOptions: boardOptionsReducer,
  currentSelectedCard: currentSelectedCardReducer,
  currentSelectedList: currentSelectedListReducer,
  listOptionsMenuPosition: listOptionsMenuPositionReducer,
  tags: tagsReducer,
  selectedTagColour: selectedTagColourReducer,
  createTagMenuData: createTagMenuDataReducer,
  modalActionMenusVisibility: modalActionMenusVisibilityReducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>
export type AppDispatch = AppStore['dispatch']
export type AppStore = ReturnType<typeof setupStore>