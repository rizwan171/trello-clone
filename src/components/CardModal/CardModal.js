import React, { useEffect, useRef, useState } from "react";
import PropTypes from "prop-types";
import { useDispatch } from "react-redux";
import { updateCardContent, deleteCard, copyCardToList, moveCardToList } from "../../features/cardsSlice";
import { clearSelectedCard, setCurrentSelectedCard } from "../../features/currentSelectedCardSlice";
import { useSelector } from "react-redux";

const CardModal = ({ card }) => {
  const dispatch = useDispatch();
  const ref = useRef();
  const lists = useSelector((state) => state.lists.value);
  const [selected, setSelected] = useState(false);
  // TODO: copy and move shouldnt be open at the same time, so these states could be reduced to just the 1
  const [copyOpen, setCopyOpen] = useState(false);
  const [moveOpen, setMoveOpen] = useState(false);
  const [tagsOpen, setTagsOpen] = useState(false);
  const [selectedListId, setSelectedListId] = useState("");
  const [editableContent, setEditableContent] = useState(card.content);
  const [rows, setRows] = useState(1);

  useEffect(() => {
    if (ref && ref.current) {
      ref.current.style.height = "0px";
      const scrollHeight = ref.current.scrollHeight;
      ref.current.style.height = scrollHeight + "px";
    }
  }, [editableContent]);

  const closeModal = () => {
    dispatch(clearSelectedCard());
  };

  const handleCardContentOnChange = (e) => {
    setEditableContent(e.target.value);
  };

  const handleCopyCardOnChange = (e) => {
    setSelectedListId(e.target.value);
  };

  const handleMoveCardOnChange = (e) => {
    setSelectedListId(e.target.value);
  };

  const handleOnBlur = () => {
    setSelected(false);
    setEditableContent(card.content);
    setRows(1);
  };

  const handleKeyDown = (e) => {
    if (e.key === "Enter") {
      setSelected(false);
      setRows(1);
      dispatch(updateCardContent({ id: card.id, content: editableContent }));
      dispatch(setCurrentSelectedCard({ ...card, content: editableContent }));
    }
  };

  const handleCopy = () => {
    if (copyOpen) {
      dispatch(copyCardToList({ cardId: card.id, listId: selectedListId }));
      setSelectedListId("");
      setCopyOpen(false);
    } else {
      setCopyOpen(true);
      setMoveOpen(false);
      setSelectedListId("");
    }
  };

  const handleMove = () => {
    if (moveOpen) {
      dispatch(moveCardToList({ cardId: card.id, listId: selectedListId }));
      setSelectedListId("");
      setMoveOpen(false);
    } else {
      setMoveOpen(true);
      setCopyOpen(false);
      setSelectedListId("");
    }
  };

  const handleDeleteCard = () => {
    dispatch(deleteCard(card.id));
    dispatch(clearSelectedCard());
  };

  const handleOnFocus = (e) => {
    // The scrollHeight is the height of all the content, including that which exceeds 1 row of textarea.
    // clientHeight is the height of the content the user can see at this point with 1 row of textarea.
    // To dynamically set the rows, we simply divide the scrollHeight by clientHeight and roundup
    setRows(Math.ceil(e.target.scrollHeight / e.target.clientHeight));
  };

  const openTagsMenu = () => {
    setTagsOpen(true);
  };

  return (
    <div
      id="cardModal"
      className="z-50 min-h-full min-w-full flex fixed top-0 left-0 justify-center items-center bg-black bg-opacity-30"
    >
      <div className="relative px-4 w-full max-w-3xl h-full md:h-auto mb-36">
        <div className="relative bg-trello-gray-100 rounded-lg shadow dark:bg-gray-700">
          <div className="flex p-6 rounded-t">
            {!selected && (
              <h3
                className="text-xl font-semibold text-gray-900 lg:text-2xl dark:text-white break-all"
                onClick={() => setSelected(true)}
              >
                {card.content}
              </h3>
            )}
            {selected && (
              <textarea
                ref={ref}
                type="text"
                value={editableContent}
                id="rounded-email"
                autoFocus
                rows={rows}
                onFocus={handleOnFocus}
                className="w-full scroll-y-hidden mr-2 py-1 text-trello-gray-300 text-xl font-semibold rounded-sm border bg-white shadow-sm focus:outline-none focus:ring-2 focus:ring-trello-blue-100"
                onBlur={handleOnBlur}
                onChange={handleCardContentOnChange}
                onKeyDown={handleKeyDown}
              />
            )}

            <button
              type="button"
              className="text-gray-400 bg-transparent hover:bg-black hover:bg-opacity-10 hover:text-gray-600 rounded-lg text-sm p-1.5 ml-auto inline-flex items-center dark:hover:bg-gray-600 dark:hover:text-white"
              onClick={closeModal}
            >
              <svg className="w-5 h-5" fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
                <path
                  fillRule="evenodd"
                  d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                  clipRule="evenodd"
                ></path>
              </svg>
            </button>
          </div>
          <div className="p-6 space-y-6">
            <div className="flex">
              <div className="w-3/4">
                <p className="text-base leading-relaxed text-gray-500 dark:text-gray-400">
                  With less than a month to go before the European Union enacts new consumer privacy laws for its citizens,
                  companies around the world are updating their terms of service agreements to comply.
                </p>
                <p className="text-base leading-relaxed text-gray-500 dark:text-gray-400">
                  The European Unionâ€™s General Data Protection Regulation (G.D.P.R.) goes into effect on May 25 and is meant to
                  ensure a common set of data rights in the European Union. It requires organizations to notify users as soon as
                  possible of high-risk data breaches that could personally affect them.
                </p>
              </div>
              <div className="flex w-1/4 flex-col">
                <h4 className="mb-2 text-gray-500 text-xs self-end">Card Options</h4>
                {tagsOpen ? (
                  <div className="flex flex-col h-48">
                    <div>Test</div>
                    <div>Test</div>
                  </div>
                ) : (
                  <button
                    className="flex w-full py-2 px-3 mb-2 bg-trello-green-100 hover:bg-trello-green-200 text-white items-center text-base shadow-md rounded-md"
                    onClick={openTagsMenu}
                  >
                    Tags
                  </button>
                )}

                {/* TODO remove cursor-not-allowed and opacity-50 after colour added to card */}
                <button className="cursor-not-allowed opacity-50 flex w-full py-2 px-3 mb-2 bg-trello-green-100 hover:bg-trello-green-200 text-white items-center text-base shadow-md rounded-md">
                  Colour
                </button>
                {moveOpen && (
                  <select
                    value={selectedListId}
                    autoFocus
                    onChange={handleMoveCardOnChange}
                    className="form-select py-2 px-3 mb-2 flex w-full text-base font-normal text-gray-700 bg-clip-padding bg-no-repeat border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
                  >
                    <option selected value="" disabled>
                      Select List...
                    </option>
                    {lists.map((list) => {
                      return (
                        <option key={list.id} value={list.id}>
                          {list.title}
                        </option>
                      );
                    })}
                  </select>
                )}
                <button
                  className="flex w-full py-2 px-3 mb-2 bg-trello-green-100 hover:bg-trello-green-200 text-white items-center text-base shadow-md rounded-md"
                  onClick={handleMove}
                >
                  {moveOpen ? "Confirm" : "Move"}
                </button>
                {copyOpen && (
                  <select
                    value={selectedListId}
                    autoFocus
                    onChange={handleCopyCardOnChange}
                    className="form-select py-2 px-3 mb-2 flex w-full text-base font-normal text-gray-700 bg-clip-padding bg-no-repeat border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
                  >
                    <option selected value="" disabled>
                      Select List...
                    </option>
                    {lists.map((list) => {
                      return (
                        <option key={list.id} value={list.id}>
                          {list.title}
                        </option>
                      );
                    })}
                  </select>
                )}
                <button
                  className="flex w-full py-2 px-3 mb-2 bg-trello-green-100 hover:bg-trello-green-200 text-white items-center text-base shadow-md rounded-md"
                  onClick={handleCopy}
                >
                  {copyOpen ? "Confirm" : "Copy"}
                </button>
                <button
                  className="flex w-full py-2 px-3 mb-2 bg-trello-green-100 hover:bg-trello-green-200 text-white items-center text-base shadow-md rounded-md"
                  onClick={handleDeleteCard}
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default CardModal;
