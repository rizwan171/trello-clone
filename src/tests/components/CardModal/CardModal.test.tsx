import { fireEvent, screen, waitFor } from "@testing-library/react";
import { RootState } from "../../../app/store";
import CardModal from "../../../components/CardModal/CardModal";
import { renderWithProviders } from "../../utils/renderUtils";

describe("CardModal", () => {
  let initialState: Partial<RootState>;

  beforeEach(() => {
    const card = {
      id: "1",
      title: "Card 1",
      listId: "1",
      description: "Description 1",
      tags: [],
      attachments: [],
    };

    initialState = {
      cards: {
        value: [card],
      },
      currentSelectedCard: {
        value: { ...card },
      },
      lists: {
        value: [
          {
            id: "1",
            title: "List 1",
          },
        ],
      },
    };
  });

  it("should render successfully", () => {
    const view = renderWithProviders(<CardModal />, { preloadedState: initialState });
    expect(view.asFragment()).toMatchSnapshot();
  });

  it("should update card title on enter press", () => {
    const { store } = renderWithProviders(<CardModal />, { preloadedState: initialState });
    expect(store.getState().cards.value[0].title).toBe("Card 1");
    const cardTitleHeader = screen.getByText("Card 1") as HTMLHeadingElement;
    expect(cardTitleHeader).toBeInTheDocument();

    fireEvent.click(cardTitleHeader);

    const cardTitleTextArea = screen.getByText("Card 1") as HTMLTextAreaElement;
    expect(cardTitleTextArea).toBeInTheDocument();
    expect(cardTitleHeader).not.toBeInTheDocument();

    fireEvent.change(cardTitleTextArea, { target: { value: "Updated Title" } });
    fireEvent.keyDown(cardTitleTextArea, { key: "Enter" });

    expect(store.getState().cards.value[0].title).toBe("Updated Title");
    expect(store.getState().currentSelectedCard.value?.title).toBe("Updated Title");
  });

  it("should not update card title on blur", () => {
    const { store } = renderWithProviders(<CardModal />, { preloadedState: initialState });
    expect(store.getState().cards.value[0].title).toBe("Card 1");
    expect(store.getState().currentSelectedCard.value?.title).toBe("Card 1");
    const cardTitleHeader = screen.getByText("Card 1") as HTMLHeadingElement;
    expect(cardTitleHeader).toBeInTheDocument();

    fireEvent.click(cardTitleHeader);

    const cardTitleTextArea = screen.getByText("Card 1") as HTMLTextAreaElement;
    expect(cardTitleTextArea).toBeInTheDocument();
    expect(cardTitleHeader).not.toBeInTheDocument();

    fireEvent.change(cardTitleTextArea, { target: { value: "Updated Title" } });
    fireEvent.blur(cardTitleTextArea);

    expect(store.getState().cards.value[0].title).toBe("Card 1");
    expect(store.getState().currentSelectedCard.value?.title).toBe("Card 1");
  });

  it("should blur card description on escape press", async () => {
    const { store } = renderWithProviders(<CardModal />, { preloadedState: initialState });
    expect(store.getState().cards.value[0].description).toBe("Description 1");
    expect(store.getState().currentSelectedCard.value?.description).toBe("Description 1");

    const cardDescriptionTextArea = screen.getByText("Description 1") as HTMLTextAreaElement;
    expect(cardDescriptionTextArea).toBeInTheDocument();

    const onBlurSpy = jest.spyOn(cardDescriptionTextArea, "blur");

    fireEvent.change(cardDescriptionTextArea, { target: { value: "Updated Description" } });
    fireEvent.keyDown(cardDescriptionTextArea, { key: "Escape" });

    expect(onBlurSpy).toHaveBeenCalled();
  });

  it("should update card description on blur", () => {
    const { store } = renderWithProviders(<CardModal />, { preloadedState: initialState });
    expect(store.getState().cards.value[0].description).toBe("Description 1");
    expect(store.getState().currentSelectedCard.value?.description).toBe("Description 1");

    const cardDescriptionTextArea = screen.getByText("Description 1") as HTMLTextAreaElement;
    expect(cardDescriptionTextArea).toBeInTheDocument();

    fireEvent.change(cardDescriptionTextArea, { target: { value: "Updated Description" } });
    fireEvent.blur(cardDescriptionTextArea);

    expect(store.getState().cards.value[0].description).toBe("Updated Description");
    expect(store.getState().currentSelectedCard.value?.description).toBe("Updated Description");
  });

  it("should close card modal on close icon clicked", () => {
    const { store } = renderWithProviders(<CardModal />, { preloadedState: initialState });
    expect(store.getState().currentSelectedCard.value).toBeDefined();

    const closeIcon = screen.getByTestId("card-modal-close-icon");
    expect(closeIcon).toBeInTheDocument();

    fireEvent.click(closeIcon);

    expect(store.getState().currentSelectedCard.value).toBeNull();
  });

  it("should close card modal on escape press", () => {
    const { store } = renderWithProviders(<CardModal />, { preloadedState: initialState });
    expect(store.getState().currentSelectedCard.value).toBeDefined();

    fireEvent.keyDown(document, { key: "Escape" });

    expect(store.getState().currentSelectedCard.value).toBeNull();
  });
});
