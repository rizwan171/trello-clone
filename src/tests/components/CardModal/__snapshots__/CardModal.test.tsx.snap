// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardModal should render successfully 1`] = `
<DocumentFragment>
  <div
    class="z-50 min-h-full min-w-full flex fixed top-0 left-0 justify-center items-center bg-black bg-opacity-30"
    id="cardModal"
  >
    <div
      class="flex flex-col bg-trello-gray-400 w-218 rounded-sm max-h-full p-4 mb-80"
    >
      <div
        class="flex"
      >
        <div
          class="flex flex-col w-full"
        >
          <div
            class="flex gap-1 items-center"
          >
            <svg
              fill="currentColor"
              height="25"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="25"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M19 3H4.99c-1.11 0-1.98.89-1.98 2L3 19c0 1.1.88 2 1.99 2H19c1.1 0 2-.9 2-2V5a2 2 0 00-2-2zm0 12h-4c0 1.66-1.35 3-3 3s-3-1.34-3-3H4.99V5H19v10z"
              />
            </svg>
            <h3
              class="text-xl font-semibold text-gray-800 break-all"
            >
              Card 1
            </h3>
          </div>
          <p
            class="text-sm text-gray-700"
          >
            in list List 1
          </p>
        </div>
        <button
          class="text-gray-400 bg-transparent hover:bg-black hover:bg-opacity-10 hover:text-gray-600 rounded-lg text-sm p-2 ml-auto inline-flex items-center dark:hover:bg-gray-600 dark:hover:text-white"
          type="button"
        >
          <svg
            data-testid="card-modal-close-icon"
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </button>
      </div>
      <div
        class="flex w-full h-full"
      >
        <div
          class="flex flex-col w-2/3 py-4"
        >
          <div
            class="flex flex-col text-sm text-gray-700"
          >
            <p
              class="mb-2"
            >
              Tags
            </p>
            <div
              class="flex gap-1 flex-wrap mb-4"
            >
              <button
                class="p-1 rounded-ibsm flex items-center cursor-pointer bg-black bg-opacity-5 hover:bg-opacity-10"
              >
                <svg
                  class="mr-0.5 text-trello-gray-200"
                  fill="none"
                  height="25"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="25"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="12"
                    x2="12"
                    y1="5"
                    y2="19"
                  />
                  <line
                    x1="5"
                    x2="19"
                    y1="12"
                    y2="12"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="flex items-center gap-2 text-lg mb-2 text-gray-800 font-semibold"
          >
            <svg
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="17"
                x2="3"
                y1="10"
                y2="10"
              />
              <line
                x1="21"
                x2="3"
                y1="6"
                y2="6"
              />
              <line
                x1="21"
                x2="3"
                y1="14"
                y2="14"
              />
              <line
                x1="17"
                x2="3"
                y1="18"
                y2="18"
              />
            </svg>
            <p>
              Description
            </p>
          </div>
          <textarea
            class="text-gray-600 p-2 bg-black bg-opacity-5 rounded-ibsm h-44"
            placeholder="Add some more detail here..."
          >
            Description 1
          </textarea>
        </div>
        <div
          class="flex flex-col w-44 py-4 ml-auto"
        >
          <h4
            class="text-gray-800 text-sm"
          >
            Add to card
          </h4>
          <div
            class="flex-col flex"
          >
            <button
              class="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
              data-testid="card-modal-actions-tags-button"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M938 458.8l-29.6-312.6c-1.5-16.2-14.4-29-30.6-30.6L565.2 86h-.4c-3.2 0-5.7 1-7.6 2.9L88.9 557.2a9.96 9.96 0 0 0 0 14.1l363.8 363.8c1.9 1.9 4.4 2.9 7.1 2.9s5.2-1 7.1-2.9l468.3-468.3c2-2.1 3-5 2.8-8zM459.7 834.7L189.3 564.3 589 164.6 836 188l23.4 247-399.7 399.7zM680 256c-48.5 0-88 39.5-88 88s39.5 88 88 88 88-39.5 88-88-39.5-88-88-88zm0 120c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32z"
                />
              </svg>
              <p>
                Tags
              </p>
            </button>
          </div>
          <h4
            class="text-gray-800 text-sm"
          >
            Actions
          </h4>
          <div
            class="flex-col flex"
          >
            <button
              class="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
              data-testid="card-modal-actions-copy-button"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0V0z"
                  fill="none"
                />
                <path
                  d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
                />
              </svg>
              <p>
                Copy
              </p>
            </button>
          </div>
          <div
            class="flex-col flex"
          >
            <button
              class="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
              data-testid="card-modal-actions-move-button"
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14 5l7 7m0 0l-7 7m7-7H3"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <p>
                Move
              </p>
            </button>
          </div>
          <div
            class="flex-col flex"
          >
            <button
              class="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
              data-testid="card-modal-actions-delete-button"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                />
              </svg>
              <p>
                Delete
              </p>
            </button>
          </div>
          <div
            class="flex-col flex"
          >
            <button
              class="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
              data-testid="card-modal-actions-attachment-menu-button"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M779.3 196.6c-94.2-94.2-247.6-94.2-341.7 0l-261 260.8c-1.7 1.7-2.6 4-2.6 6.4s.9 4.7 2.6 6.4l36.9 36.9a9 9 0 0 0 12.7 0l261-260.8c32.4-32.4 75.5-50.2 121.3-50.2s88.9 17.8 121.2 50.2c32.4 32.4 50.2 75.5 50.2 121.2 0 45.8-17.8 88.8-50.2 121.2l-266 265.9-43.1 43.1c-40.3 40.3-105.8 40.3-146.1 0-19.5-19.5-30.2-45.4-30.2-73s10.7-53.5 30.2-73l263.9-263.8c6.7-6.6 15.5-10.3 24.9-10.3h.1c9.4 0 18.1 3.7 24.7 10.3 6.7 6.7 10.3 15.5 10.3 24.9 0 9.3-3.7 18.1-10.3 24.7L372.4 653c-1.7 1.7-2.6 4-2.6 6.4s.9 4.7 2.6 6.4l36.9 36.9a9 9 0 0 0 12.7 0l215.6-215.6c19.9-19.9 30.8-46.3 30.8-74.4s-11-54.6-30.8-74.4c-41.1-41.1-107.9-41-149 0L463 364 224.8 602.1A172.22 172.22 0 0 0 174 724.8c0 46.3 18.1 89.8 50.8 122.5 33.9 33.8 78.3 50.7 122.7 50.7 44.4 0 88.8-16.9 122.6-50.7l309.2-309C824.8 492.7 850 432 850 367.5c.1-64.6-25.1-125.3-70.7-170.9z"
                />
              </svg>
              <p>
                Attachments
              </p>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
